tokio/src/scope/cancellation_token.rs

```diff
@@ -0,0 +1,1186 @@
+                    };
+                }
+                guard.first_child = Some((&mut *child_token_state).into());
+            }
+        };
+
+        let child_token_ptr = Box::into_raw(child_token_state);
+        // Safety: We just created the pointer from a `Box`
+        CancellationToken {
+            inner: unsafe { NonNull::new_unchecked(child_token_ptr) },
+        }
+    }
+
+    /// Returns the number of child tokens
+    #[cfg(all(test, not(loom)))]
+    fn child_tokens(&self) -> usize {
+        let mut result = 0;
+        let inner = self.state();
+
+        let guard = inner.synchronized.lock().unwrap();
+        let mut child = guard.first_child;
+        while let Some(mut c) = child {
+            result += 1;
+            // Safety: The child state is accessed from within a Mutex. Since
+            // the child needs to take the Mutex to unregister itself before
+            // getting destroyed, it is guaranteed to be alive.
+            child = unsafe { c.as_mut().from_parent.next_peer };
+        }
+
+        result
+    }
+
+    /// Cancel the [`CancellationToken`] and all child tokens which had been
+    /// derived from it.
+    ///
+    /// This will wake up all tasks which are waiting for cancellation.
+    #[allow(dead_code)]
+    pub(crate) fn cancel(&self) {
+        self.state().cancel();
+    }
+
+    /// Returns `true` if the `CancellationToken` had been cancelled
+    pub fn is_cancelled(&self) -> bool {
+        self.state().is_cancelled()
+    }
+
+    /// Returns a `Future` that gets fulfilled when cancellation is signalled.
+    pub fn wait_for_cancellation(&self) -> WaitForCancellationFuture<'_> {
+        WaitForCancellationFuture {
+            cancellation_token: Some(self),
+            wait_node: ListNode::new(WaitQueueEntry::new()),
+            is_registered: false,
+        }
+    }
+
+    unsafe fn register(
+        &self,
+        wait_node: &mut ListNode<WaitQueueEntry>,
+        cx: &mut Context<'_>,
+    ) -> Poll<()> {
+        self.state().register(wait_node, cx)
+    }
+
+    fn check_for_cancellation(
+        &self,
+        wait_node: &mut ListNode<WaitQueueEntry>,
+        cx: &mut Context<'_>,
+    ) -> Poll<()> {
+        self.state().check_for_cancellation(wait_node, cx)
+    }
+
+    fn unregister(&self, wait_node: &mut ListNode<WaitQueueEntry>) {
+        self.state().unregister(wait_node)
+    }
+}
+
+/// A Future that is resolved once the corresponding [`CancellationToken`]
+/// was cancelled
+#[must_use = "futures do nothing unless polled"]
+pub struct WaitForCancellationFuture<'a> {
+    /// The CancellationToken that is associated with this WaitForCancellationFuture
+    cancellation_token: Option<&'a CancellationToken>,
+    /// Node for waiting at the cancellation_token
+    wait_node: ListNode<WaitQueueEntry>,
+    /// Whether this future was registered at the token yet as a waiter
+    is_registered: bool,
+}
```

Comment:
I'm not opposed to exposing the future in the public API, but to start out, we _may_ want to just make this private and change the function that returning it to an `async fn` that awaits this future? That decreases the size of the public API

Reply:
As long as we could still `select!` on it that should be fine

Reply:
So when trying to move things to `pub async fn cancelled(&self)` I found out that calling this would return a `Future` which is not `Send`, whereas the current `WaitForCancellation` Future is. Afaik for some reason in `scope` this was important, but I don't remember anymore.

Another option is obviously to return `impl Future<Output=()> + Send`, but I'm not sure whether this is preferrable. We can also try starting with `async fn` and adding the concrete type back if `Send` is indeed required.

Reply:
Hmm...I think having the actual future type is fine? Elsewhere in Tokio, we've been trying to provide `async fn`-based APIs where possible, but i think being `Send` is important, due to spawning. I'd get @carllerche's opinion?